libdnf/dnf-swdb-groups.c

```diff
@@ -0,0 +1,966 @@
+                                            (gchar *)sqlite3_column_text (res, 3),       // ui_name
+                                            sqlite3_column_int (res, 4), // pkg_types
+                                            sqlite3_column_int (res, 5), // grp_types
+                                            self);                       // swdb
+        env->eid = sqlite3_column_int (res, 0);
+        g_ptr_array_add (node, (gpointer) env);
+    }
+    sqlite3_finalize (res);
+    for (guint i = 0; i < node->len; ++i) {
+        DnfSwdbEnv *env = g_ptr_array_index (node, i);
+        _env_installed (self->db, env);
+    }
+    return node;
+}
+
+/**
+ * _get_list_from_table:
+ * @res: sqlite3_stmt returning list of strings
+ *
+ * Returns: list of string matched by @res
+ **/
+GPtrArray *
+_get_list_from_table (sqlite3_stmt *res)
+{
+    GPtrArray *node = g_ptr_array_new ();
+    while (sqlite3_step (res) == SQLITE_ROW) {
+        gchar *str = g_strdup ((const gchar *)sqlite3_column_text (res, 0));
+        g_ptr_array_add (node, (gpointer)str);
+    }
+    sqlite3_finalize (res);
+    return node;
+}
+
+/**
+ * _get_data_list:
+ * @db: sqlite database handle
+ * @gid: group id
+ * @table: target table
+ *
+ * Get list of names connected with @gid from @table
+ *
+ * Returns: list of names
+ *
+ * TODO: don't use dynamic sql command construction
+ **/
+GPtrArray *
+_get_data_list (sqlite3 *db, int gid, const gchar *table)
+{
+    sqlite3_stmt *res;
+    g_autofree gchar *sql = g_strjoin (" ", "SELECT name FROM", table, "where G_ID=@gid", NULL);
+    DB_PREP (db, sql, res);
+    DB_BIND_INT (res, "@gid", gid);
+    GPtrArray *node = _get_list_from_table (res);
+    return node;
+}
+
+/**
+ * dnf_swdb_group_get_exclude:
+ * @self: group object
+ *
+ * Get list of excludes for @self
+ *
+ * Returns: (element-type utf8)(array)(transfer container): list of excludes
+ **/
+GPtrArray *
+dnf_swdb_group_get_exclude (DnfSwdbGroup *self)
+{
+    if (!self->gid)
+        return NULL;
+    if (dnf_swdb_open (self->swdb))
+        return NULL;
+    GPtrArray *node = _get_data_list (self->swdb->db, self->gid, "GROUPS_EXCLUDE");
+    return node;
+}
+
+/**
+ * dnf_swdb_group_get_full_list:
+ * @self: group object
+ *
+ * Get full package list of @self
+ *
+ * Returns: (element-type utf8)(array)(transfer container): list of package names
+ **/
+GPtrArray *
+dnf_swdb_group_get_full_list (DnfSwdbGroup *self)
+{
+    if (!self->gid)
+        return NULL;
+    if (dnf_swdb_open (self->swdb))
+        return NULL;
+    GPtrArray *node = _get_data_list (self->swdb->db, self->gid, "GROUPS_PACKAGE");
+    return node;
+}
+
+/**
+ * dnf_swdb_group_update_full_list:
+ * @group: group object
+ * @full_list: (element-type utf8)(transfer container): list of package names
+ *
+ * Update list of packages installed with group
+ *
+ * Returns: 0 if successful
+ */
+gint
+dnf_swdb_group_update_full_list (DnfSwdbGroup *group, GPtrArray *full_list)
```

Comment:
Function name can be more specific. What is meaning of "update"? setting, inserting, deleting? I must look to source of function to see that it is inserting. And better description also helps.

Reply:
**const** GPtrArray *full_list